formats/json-tests/commonTest/src/kotlinx/serialization/json/polymorphic/JsonTreeDecoderPolymorphicTest.kt

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2017-2023 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.json.polymorphic
+
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.json.*
+import kotlin.test.*
+
+class JsonTreeDecoderPolymorphicTest {
+
+    @Serializable
+    sealed class Sealed
+
+    @Serializable
+    data class ClassContainingItself(
+        val a: String,
+        val b: String,
+        val c: ClassContainingItself? = null,
+        val d: String?
+    ) : Sealed()
+
+    val inner = ClassContainingItself(
+        "InnerA",
+        "InnerB",
+        null,
+        "InnerC"
+    )
+    val outer = ClassContainingItself(
+        "OuterA",
+        "OuterB",
+        inner,
+        "OuterC"
+    )
+
+    @Test
+    fun testDecodingWhenClassContainsItself() {
```

Comment:
Please use `JsonTestBase` as a base class for test and function `parametrizedTest`. It helps check behavior for all possible json decoding paths, including JsonTreeDecoder (problematic one), as well as String decoder and Stream integration. You can see example of using one here: https://github.com/Kotlin/kotlinx.serialization/blob/5e8ccad1f70a9457e0ffe6ae6b10a0bd0eaaa618/formats/json-tests/commonTest/src/kotlinx/serialization/json/JsonReifiedCollectionsTest.kt#L15